skimage/data/__init__.py

```diff
@@ -49,6 +54,159 @@
            'shepp_logan_phantom',
            'stereo_motorcycle']
 
+legacy_data_dir = osp.abspath(osp.dirname(__file__))
+skimage_distribution_dir = osp.join(legacy_data_dir, '..')
+
+try:
+    import pooch
+    from pooch.utils import file_hash
+    has_pooch = True
+except ImportError:
+    has_pooch = False
+    import hashlib
+
+    def file_hash(fname):
+        # Versioned from pooch.utils
+        # Calculate the hash in chunks to avoid overloading the memory
+        chunksize = 65536
+        hasher = hashlib.sha256()
+        with open(fname, "rb") as fin:
+            buff = fin.read(chunksize)
+            while buff:
+                hasher.update(buff)
+                buff = fin.read(chunksize)
+        return hasher.hexdigest()
+
+
+def _fetch(data_filename):
+    # potential_local_file is the location of the file that would be shipped
+    # with scikit-image.
+    #
+    # Two things can cause the file to already be in the distribution:
+    # 1. The user is running it from the github repo, in which case
+    #    we should avoid a downloading from the internet
+    # 2. We chose to ship this file with skimage by default
+    potential_local_file = osp.join(skimage_distribution_dir, data_filename)
+    resolved_path = osp.join(data_dir, '..', data_filename)
+    expected_hash = registry[data_filename]
+    if osp.exists(resolved_path) and file_hash(resolved_path) == expected_hash:
+        # Nothing to be done, file is where it is expected to be
+        pass
+    elif (osp.exists(potential_local_file)
+            and file_hash(potential_local_file) == expected_hash):
+        parent = osp.dirname(resolved_path)
+        os.makedirs(parent, exist_ok=True)
+        shutil.copy2(potential_local_file, resolved_path)
+    elif has_pooch:
+        resolved_path = image_fetcher.fetch(data_filename)
+    elif data_filename in registry:
+        # Legacy data might be available even without pooch
+        raise RuntimeError(
+            "Loading the file {data_filename} requires the dependency "
+            "'pooch' to be installed. Install it with the command "
+            "'pip install pooch' or 'conda install pooch' as appropriate."
+            "".format(data_filename=data_filename))
+    else:
+        # This is only expected to happen if the user calls
+        # ``skimage.data.load`` with a filename that isn't part of the
+        # skimage distribution
+        raise ValueError(
+            "The resquested file {data_filename} is not part of the "
+            "scikit-image distribution."
+            "".format(data_filename=data_filename))
+    return resolved_path
+
+
+if has_pooch:
+    # Pooch expects a `+` to exist in development versions.
+    # Since scikit-image doesn't follow that convention, we have to manually
+    # remove `.dev` with a `+` if it exists.
+    # This helps pooch understand that it should look in master
+    # to find the required files
+    pooch_version = __version__.replace('.dev', '+')
```

Comment:
I'm confused now: are the image files going to live in this repository, or elsewhere?

Reply:
The image files **can** be added to this repository, but they will no longer be added to the the pypi distribution.

I think we should avoid adding more binary files to the github repo, since this will enable us to add much larger files elsewhere.